charms/worker/k8s/src/certificates.py

```diff
@@ -0,0 +1,280 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""K8s Certificates module."""
+import logging
+from typing import List
+
+import ops
+import utils
+from charms.k8s.v0.k8sd_api_manager import BootstrapConfig
+from charms.tls_certificates_interface.v3.tls_certificates import (
+    CertificateAvailableEvent,
+    generate_csr,
+    generate_private_key,
+)
+
+log = logging.getLogger(__name__)
+
+WORKER_COMPONENTS = ["kubelet"]
+CONTROL_PLANE_COMPONENTS = ["apiserver", "front-proxy-client"] + WORKER_COMPONENTS
+LEADER_CONTROL_PLANE_COMPONENTS = ["apiserver-kubelet-client"] + CONTROL_PLANE_COMPONENTS
+
+
+class K8sCertificates:
+    """A class for managing Kubernetes certificates associated with a cluster unit."""
+
+    def __init__(self, charm, certificates) -> None:
+        """Initialize the K8sCertificates class.
+
+        Args:
+            charm: An instance of the charm.
+            certificates: An instance of the TLSCertificatesv3 library.
+        """
+        self.charm = charm
+        self.certificates = certificates
+
+    def _check_secret(self, label: str):
+        """Check if a secret with the specified label exists in Juju.
+
+        Args:
+            label (str): The label of the secret to check for.
+
+        Returns:
+            bool: True if the secret exists, False otherwise.
+        """
+        try:
+            self.charm.model.get_secret(label=label)
+            return True
+        except ops.SecretNotFoundError:
+            return False
+
+    def _generate_private_key(self, components: List[str]):
+        """Generate private keys for each component and store them as secrets.
+
+        Args:
+            components (List[str]): A list of components to generate private keys for.
+        """
+        content = dict()
+
+        for key in components:
+            private_key = generate_private_key()
+            content[key] = private_key.decode()
+
+        self.charm.unit.add_secret(content=content, label=f"{self.charm.unit.name}-private-keys")
```

Comment:
Does it make sense that these csr/private-keys would ever need rotation?  Should they have secrets that expire and need to be regenerated?

Reply:
I don't think so. When we receive an event that signals an expiring certificate, we need to provide both the old CSR and the private key to attempt the renewal. For instance:
```python
    def _on_certificate_expiring(
        self, event: Union[CertificateExpiringEvent, CertificateInvalidatedEvent]
    ) -> None:
        replicas_relation = self.model.get_relation("replicas")
        if not replicas_relation:
            self.unit.status = WaitingStatus("Waiting for peer relation to be created")
            event.defer()
            return
        old_csr = replicas_relation.data[self.app].get("csr")
        private_key_password = replicas_relation.data[self.app].get("private_key_password")
        private_key = replicas_relation.data[self.app].get("private_key")
        new_csr = generate_csr(
            private_key=private_key.encode(),
            private_key_password=private_key_password.encode(),
            subject=self.cert_subject,
        )
        self.certificates.request_certificate_renewal(
            old_certificate_signing_request=old_csr,
            new_certificate_signing_request=new_csr,
        )
        replicas_relation.data[self.app].update({"csr": new_csr.decode()})
```
Seeing this usage from the library examples, I think we need to keep both CSR and Private Keys to ease the certificate management.